contracts/src/utils/nonces.rs

```diff
@@ -60,13 +61,19 @@ impl Nonces {
     /// * `&mut self` - Write access to the contract's state.
     /// * `owner` - The address for which to consume the nonce.
     ///
-    /// # Panics
+    /// # Safety
     ///
-    /// This function will panic if the nonce for the given `owner` has reached
-    /// the maximum value representable by `U256`, causing the `checked_add`
-    /// method to return `None`.
+    /// the function will experience [undefined behavior], if the nonce for
+    /// the given `owner` has reached the maximum value representable by `U256`.
+    /// Extreme caution should be taken to ensure overflow cannot occur.
+    ///
+    /// [undefined behavior]: https://doc.rust-lang.org/reference/behavior-considered-undefined.html
     pub fn use_nonce(&mut self, owner: Address) -> U256 {
         let nonce = self._nonces.get(owner);
+
+        // For each account, the nonce has an initial value of 0, can only be
+        // incremented by one, and cannot be decremented or reset. This
+        // guarantees that the nonce never overflows.
         self._nonces
             .setter(owner)
             .set(unsafe { nonce.checked_add(ONE).unwrap_unchecked() });
```

Comment:
Tbh I do not like code where we use `unsafe` in our contracts. Is it possible to apply panicking operation before the setter?

Reply:
I don't really understand how this "guarantees that the nonce never overflows", but that's what's in the Solidity version.

How bad performance-wise can it be if we _did_ check for overflows here?

Reply:
I believe we should have panic here to make it obvious that it may fail, like:
```rust
let updated_nonce = nonce.checked_add(ONE).expect("...");
 self._nonces
            .setter(owner)
            .set(updated_nonce);
```

Reply:
And then we need to have `# Panics` sections in Rust docs for proper fns.

Reply:
@bidzyyys @qalisander I think this was mentioned before, but it would be useful to have "checked" add/sub assign versions in [math/storage.rs](https://github.com/OpenZeppelin/rust-contracts-stylus/blob/59c1ead477e969e0d85b27ac8784c760acbb8ca0/contracts/src/utils/math/storage.rs):

```rust
trait AddAssignChecked
trait SubAssignChecked
```

Reply:
@0xNeshi yeah, having "safe" (`Checked`) operation would be great. If you want to include those traits in this PR go ahead; If not feel free to add an issue to this milestone.

Reply:
I'll add a separate issue, as there are many contracts where the change could be applied

Reply:
Here's the issue https://github.com/OpenZeppelin/rust-contracts-stylus/issues/469